src/odf/dataset-impl/src/entities/metadata_chain_validators.rs

```diff
@@ -856,6 +856,47 @@ impl<'a> ValidateSetDataSchemaVisitor<'a> {
             vocab: None,
         })
     }
+
+    fn check_allowed_schema_migration(diff: &DataSchemaDiff<'_>) -> bool {
+        diff.items
+            .iter()
+            .all(Self::check_allowed_schema_migration_rec)
+    }
+
+    fn check_allowed_schema_migration_rec(diff: &DataSchemaDiffItem<'_>) -> bool {
+        #[expect(clippy::match_same_arms)]
+        match diff {
+            DataSchemaDiffItem::SchemaAttributesChanged => true,
+            DataSchemaDiffItem::FieldAttributesChanged => true,
+            DataSchemaDiffItem::FieldAdded { field_rhs } => {
+                if field_rhs.is_optional() {
+                    // It's OK to introduce a new optional field as this maintains compatibility
+                    // from publisher and consumer perspective
+                    true
+                } else {
+                    false
+                }
+            }
+            DataSchemaDiffItem::FieldRemoved => false,
+            DataSchemaDiffItem::FieldReordered {
+                index_lhs: _,
+                index_rhs: _,
+                field_lhs: _,
+                field_rhs: _,
+            } => true,
+            DataSchemaDiffItem::FieldTypeChanged => false,
+            // TODO: Field becoming optional is relaxing the schema and a compatibile change from
+            // producer perspective, but it's a breaking change from consumer perspective and needs
+            // to be cautioned about and communicated downstream
+            DataSchemaDiffItem::FieldBecameOptional => false,
+            DataSchemaDiffItem::FieldBecameRequired => false,
```

Comment:
Field becoming required does not sound like a breaking change from consumers point of view.